internal/controller/apiextensions/composite/reconciler.go

```diff
@@ -240,41 +218,52 @@ func WithConnectionPublisher(p ConnectionPublisher) ReconcilerOption {
 type compositeResource struct {
 	CompositionSelector
 	Configurator
+	Composer
 	ConnectionPublisher
 }
 
-type composedResource struct {
-	Renderer
-	ConnectionDetailsFetcher
-	ReadinessChecker
-}
-
 // NewReconciler returns a new Reconciler of composite resources.
 func NewReconciler(mgr manager.Manager, of resource.CompositeKind, opts ...ReconcilerOption) *Reconciler {
 	nc := func() resource.Composite {
 		return composite.New(composite.WithGroupVersionKind(schema.GroupVersionKind(of)))
 	}
 	kube := unstructured.NewClient(mgr.GetClient())
 
+	composed := composedResource{
+		Renderer:                 NewAPIDryRunRenderer(kube),
+		ReadinessChecker:         ReadinessCheckerFn(IsReady),
+		ConnectionDetailsFetcher: NewAPIConnectionDetailsFetcher(kube),
+	}
+
 	r := &Reconciler{
-		client: resource.ClientApplicator{
-			Client:     kube,
-			Applicator: resource.NewAPIPatchingApplicator(kube),
-		},
+		client:       kube,
 		newComposite: nc,
 
+		composition: ValidationChain{
```

Comment:
The current implementation uses the array index as template name. What do you think about using one mutator instead of the validators that will update the `Composition` resource with templates having their index as `templateName` if it's `nil`? We'd not need the `AnonymousComposer` since all will be guaranteed to have a `templateName`.